### PR TITLE
[SYCL][CUDA][libclc] Add atomic loads and stores with various memory orders and scopes

### PR DESCRIPTION
**Pull request by t4c1**
Adds atomic loads and stores with various memory orders and scopes. These are implemented by adding intrinsics and builtins for   PTX loads and stores.

Tests for this are here: https://github.com/intel/llvm-test-suite/pull/648

This blocks https://github.com/intel/llvm/pull/4853 (but it is not the only blocker).

### CODE REVIEW

**Comment by t4c1** [comment]
This test failure does not look connected to the changes in this PR. Also, I can not reproduce it locally.

**Comment by bader** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/648

File: libclc/ptx-nvidiacl/libspirv/atomic/atomic_load.cl
Comment (bader) [inline libclc/ptx-nvidiacl/libspirv/atomic/atomic_load.cl:56]: This macro requires additional formatting. Please, apply clang-format.

**Review by bader — APPROVED** [review]
`libclc` part looks good to me.
It would be good to review LLVM intrinsics with @Artem-B.
You can either create another review request on reviews.llvm.org or extend https://reviews.llvm.org/D112718 and ping reviewers.

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1808]: Would it make sense to always explicitly use `gpu` suffix for the builtins/intrinsics?

File: clang/test/CodeGen/builtins-nvptx.c
Comment (Artem-B) [inline clang/test/CodeGen/builtins-nvptx.c:300]: Interesting. The builtins are declared with the pointer in the generic AS, yet here we're passing a pointer to the global AS and it does get us the right intrinsic overload. This is convenient. 

The downside of this is that unless user explicitly casts the pointer to a specific AS, we'll likely have unnecessary ASC to the generic AS, as I've mentioned above. It's not a bug as accesses via generic AS are OK, but rather an optimization opportunity.

File: clang/test/CodeGen/builtins-nvptx.c
Comment (Artem-B) [inline clang/test/CodeGen/builtins-nvptx.c:298]: It would be great to attempt actually assembling PTX we generate for the tests.
In the past it helped a lot to catch cases where we were generating something that `ptxas` would not accept.

**Review by Artem-B — COMMENTED** [review]
Looks OK to me, modulo few nits.

Do you plan to upstream these changes?

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (t4c1) [reply clang/include/clang/Basic/BuiltinsNVPTX.def:2048]: That would be nice to have, but I would argue, that being consistent with other builtins is more important. I was looking especially at atomic read-modify-write operations I worked on before, where we have for example `__nvvm_atom_acq_rel_sys_min_shared_i`.

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td
Comment (t4c1) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1808]: I think we should follow the pattern of existing builtins/intrinsics, which do not explicitly specify `gpu`.

File: clang/test/CodeGen/builtins-nvptx.c
Comment (t4c1) [reply clang/test/CodeGen/builtins-nvptx.c:298]: This code is also tested in llvm-test-suite, where it is compiled, assembled and run. See https://github.com/intel/llvm-test-suite/pull/648

File: clang/test/CodeGen/builtins-nvptx.c
Comment (t4c1) [reply clang/test/CodeGen/builtins-nvptx.c:300]: I am not sure I follow your thinking. This is lowered into a PTX instruction that uses global (not generic) addressing. Are you saying that compiler will cast this pointer to generic AS and than back to global AS? Also, what does ASC stand for?

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (t4c1) [reply clang/include/clang/Basic/BuiltinsNVPTX.def:2117]: done

File: libclc/ptx-nvidiacl/libspirv/atomic/atomic_load.cl
Comment (t4c1) [reply libclc/ptx-nvidiacl/libspirv/atomic/atomic_load.cl]: done

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (FMarno) [inline clang/lib/CodeGen/CGBuiltin.cpp:57]: I don't believe this is used in your changes

File: clang/test/CodeGen/builtins-nvptx.c
Comment (bader) [reply clang/test/CodeGen/builtins-nvptx.c:298]: If "assembling PTX" doesn't require NV HW, so we can put such tests to clang/test/*, it's better than running SYCL application.
Running SYCL application is more expensive as it requires NV HW, whereas clang/test/* allow developers to catch bugs much earlier in their development pipeline.

**Comment by bader** [comment]
> Do you plan to upstream these changes?

I think it would be great to contribute these changes to llvm-project.
LLVM/Clang changes can be committed right away, but libclc changes rely on other work, which should be upstreamed first.
This is the second patch adding NVPTX intrinsics. First patch https://reviews.llvm.org/D112718 is already up for review and we would appreciate your help with it (it's been waiting for almost two months).

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (Artem-B) [reply clang/include/clang/Basic/BuiltinsNVPTX.def:2050]: OK. We can live with the order of modifiers matching that of `__nvvm_atom`.

That said, `__nvvm_atom` does have the instruction mnemonic `atom` in front and so should these `ld` builtins.

File: llvm/lib/Target/NVPTX/NVPTXIntrinsics.td
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXIntrinsics.td:1808]: OK.

File: clang/test/CodeGen/builtins-nvptx.c
Comment (Artem-B) [reply clang/test/CodeGen/builtins-nvptx.c:300]: > I am not sure I follow your thinking. 

My fault. I should not do code reviews at 1am. Please ignore the second paragraph.

> This is lowered into a PTX instruction that uses global (not generic) addressing. 

This is the "interesting" part. Because the argument has been explicitly marked as being in global AS, things do work exactly as intended. In a way you've created an equivalent of a function overload for the builtin for pointers in different AS and that may allow clang to accept more than it should.

> Are you saying that compiler will cast this pointer to generic AS and than back to global AS? 

Normally, yes. See https://godbolt.org/z/aG1bd45ez. 

For the calls of `load` the pointer gets cast to global AS and then back to generic, before passing it to a function. You can see it in the IR, as well. Here you're probably saved by the fact that you handle call generation yourself, but you do sort of break the rules in the process. I.e. builtins with generic AS pointer arguments should not accept pointers in the specific AS. The test above demonstrates that these builtins do.

I'm curious what's going to happen if you actually call `__nvvm_volatile_ld_global_i(ip);`.
Technically, it does match the builtin signature, so clang will accept this. The way builtin code gets called, it will likely end up calling `@llvm.nvvm.ld.global.i.volatile.i32.p0i32`  that would probably translate into `ld.volatile` instruction with a generic pointer as an argument. I do not think it's a desired outcome.

I think builtins should have signatures with the pointers in the matching AS. This way clang will catch attempts to pass a wrong-AS pointer to them.

> Also, what does ASC stand for?
`ASC` = `addrspacecast` LLVM IR instruction.

File: clang/test/CodeGen/builtins-nvptx.c
Comment (Artem-B) [reply clang/test/CodeGen/builtins-nvptx.c:298]: > If "assembling PTX" doesn't require NV HW, so we can put such tests to clang/test/*

We'll still need `ptxas` so such test can't be part of LLVM/clang's own tests. Test-suite is the right place.

**Comment by Artem-B** [comment]
> First patch https://reviews.llvm.org/D112718 is already up for review and we would appreciate your help with it (it's been waiting for almost two months).

I do not think I've seen it. In the future, please do ping the review and explicitly mention reviewers if there's no activity. There are a lot of review email flying around and sometimes things do fall through the cracks. I'm off for the holidays, but I'll take a look when I'm back.

File: llvm/test/CodeGen/NVPTX/atomics-with-semantics.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/atomics-with-semantics.ll:2129]: Are there tests for `ld.volatile` variants?

File: clang/include/clang/Basic/BuiltinsNVPTX.def
Comment (t4c1) [reply clang/include/clang/Basic/BuiltinsNVPTX.def:2050]: My understanding is that these have `atom` at the start because they also have `atom` in the PTX, which is not true for `ld` and `st` instructions.

File: clang/test/CodeGen/builtins-nvptx.c
Comment (t4c1) [reply clang/test/CodeGen/builtins-nvptx.c:300]: > For the calls of load the pointer gets cast to global AS and then back to generic, before passing it to a function.

I don't see any such casts to global and back to generic in generated PTX. That is either in the godbolt example you linked or locally when I tried something similar with the new builtins.

> I'm curious what's going to happen if you actually call __nvvm_volatile_ld_global_i(ip);.
Technically, it does match the builtin signature, so clang will accept this. The way builtin code gets called, it will likely end up calling @llvm.nvvm.ld.global.i.volatile.i32.p0i32 that would probably translate into ld.volatile instruction with a generic pointer as an argument. I do not think it's a desired outcome.

That is exactly what happens.

> I think builtins should have signatures with the pointers in the matching AS. This way clang will catch attempts to pass a wrong-AS pointer to them.

I can not easily specify which AS the argument must be in when declaring builtins. While setting the address space numbers in builtin declarations works for CUDA, it does not work for OpenCL, which is our use case for this. I could try to throw an error in code generation if a builtin is called with AS that is not the same one it accepts.

File: llvm/test/CodeGen/NVPTX/atomics-with-semantics.ll
Comment (t4c1) [reply llvm/test/CodeGen/NVPTX/atomics-with-semantics.ll:2129]: They were missing. I added them now.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (t4c1) [reply clang/lib/CodeGen/CGBuiltin.cpp]: removed

File: clang/test/CodeGen/builtins-nvptx.c
Comment (t4c1) [reply clang/test/CodeGen/builtins-nvptx.c:300]: I would suggest that we merge this as it is to unblock other atomic-related PRs. 

When we figure out how to enable typecasting between target and OpenCL language address spaces we come back to this and change the builtins in the way that was suggested in this thread. Regardless of what we decide here this will need to be done for atomic read-modify-write builtins that have already been merged.

Comment (bader) [reply clang/test/CodeGen/builtins-nvptx.c:300]: I'm okay with that plan if @Artem-B has no objections.

File: clang/lib/CodeGen/CGBuiltin.cpp
Comment (elizabethandrews) [inline clang/lib/CodeGen/CGBuiltin.cpp:17434]: You can return this directly. `tmp` is not required

**Review by elizabethandrews — COMMENTED** [review]
Just commented a nit. 

I am not familiar with this functionality and so I leave the review to @bader.

**Review by bader — DISMISSED** [review]

**Review by elizabethandrews — APPROVED** [review]
Approving to unblock since code looks ok to me. But please wait for approval from @bader or someone else who is familiar with this functionality

**Review by smanna12 — APPROVED** [review]

**Comment by bader** [comment]
@t4c1, I'm lost a bit in PR dependencies, so I need your help here to avoid merging something that will break CI checks.
- Is this PR breaks llvm-test-suite tests, which are supposed to be fixed by https://github.com/intel/llvm-test-suite/pull/648?
- https://github.com/intel/llvm-test-suite/pull/648 says it's blocked by https://github.com/intel/llvm/pull/4853, which is a draft and can't be merged. If we need to merge https://github.com/intel/llvm-test-suite/pull/648 together with this PR, we need to merge https://github.com/intel/llvm/pull/4853 first. Is that right?

**Comment by t4c1** [comment]
Merging this will not break anything. 

https://github.com/intel/llvm-test-suite/pull/648 just tests new features added by this PR and does indeed need https://github.com/intel/llvm/pull/4853 merged first.

However this PR is the last one that blocks https://github.com/intel/llvm/pull/4853, so that one can be merged after this one is.

To summarize: atomic PRs should be merged in the order:
 - this one first 
 - than https://github.com/intel/llvm/pull/4853
 - Lastly https://github.com/intel/llvm-test-suite/pull/648 and https://github.com/intel/llvm-test-suite/pull/534 in any order

**Pull request merged**
